src/openrct2/interface/Viewport.cpp

```diff
@@ -1774,26 +1760,25 @@ static rct_viewport* viewport_find_from_point(ScreenCoordsXY screenCoords)
  */
 CoordsXY screen_get_map_xy(ScreenCoordsXY screenCoords, rct_viewport** viewport)
 {
-    int16_t my_x, my_y;
     int32_t interactionType;
     rct_viewport* myViewport = nullptr;
+    CoordsXY map_pos;
     get_map_coordinates_from_pos(
-        screenCoords.x, screenCoords.y, VIEWPORT_INTERACTION_MASK_TERRAIN, &my_x, &my_y, &interactionType, nullptr,
-        &myViewport);
+        screenCoords, VIEWPORT_INTERACTION_MASK_TERRAIN, map_pos, &interactionType, nullptr, &myViewport);
     if (interactionType == VIEWPORT_INTERACTION_ITEM_NONE)
     {
         return { LOCATION_NULL, 0 };
     }
 
     LocationXY16 start_vp_pos = screen_coord_to_viewport_coord(myViewport, screenCoords);
-    CoordsXY map_pos = { my_x + 16, my_y + 16 };
+    map_pos = { map_pos.x + 16, map_pos.y + 16 };
 
     for (int32_t i = 0; i < 5; i++)
     {
         int32_t z = tile_element_height(map_pos);
         map_pos = viewport_coord_to_map_coord(start_vp_pos.x, start_vp_pos.y, z);
-        map_pos.x = std::clamp<int16_t>(map_pos.x, my_x, my_x + 31);
-        map_pos.y = std::clamp<int16_t>(map_pos.y, my_y, my_y + 31);
+        map_pos.x = std::clamp<int32_t>(map_pos.x, map_pos.x, map_pos.x + 31);
```

Comment:
This is subtly different. my_x and my_y are the initial position. map_pos is the new position. map_pos is initialised with a value that will round up to the nearest tile.

Reply:
Nice catch!